### DISCUSSION TITLE
save video with optimizations

### DISCUSSION THREAD

**Discussion by Katsumoto7** [bloc97/Anime4K#203] [General]
Just one question, is there any way to save the video I'm watching in mpv.exe with shader optimization?

**Comment by christoph-heinrich** [comment]
mpv can encode video and I'd assume that includes the changes from the shader (never tried it), but at that point you could just directly use ffmpeg. It can use shaders via libplacebo.

**Reply by Jules-A** [reply]
ffmpeg needs a lot of VRAM, it can do some files fine but most end up failing running out of VRAM with 12GB... I think the MPV method requires Linux?

**Reply by Albidoten** [reply]
bruh can you tell me how to do that in mpv

**Reply by christoph-heinrich** [reply]
https://mpv.io/manual/master/#encoding

**Reply by Albidoten** [reply]
it shows this whenever i write command
![image](https://github.com/bloc97/Anime4K/assets/159302410/6b1219e9-1a67-403d-8913-63fcd4f2475b)

**Reply by christoph-heinrich** [reply]
You need to provide that as an argument when starting mpv from the terminal.

You can probably also do that in the console with `set o <path_to_output_file>` (I never tried that)

**Reply by Albidoten** [reply]
as a newbie in this, what should i master, so that my mind not explode by the multiple failures that i've seen from last two days and be worth the time i spent doing this cause i think the reason of my utter failure is that i know nothing about it 
![image](https://github.com/bloc97/Anime4K/assets/159302410/7cc93c1a-5ecb-47e4-8103-15781a9a2186)

i can't even tell if i am doing something wrong

**Reply by christoph-heinrich** [reply]
The simplest case should be something like `mpv --o=1.mkv D:\Video\whatever.mp4`, although on windows it might be `mpv.com` instead of `mpv`.

**Reply by Albidoten** [reply]
bro just tell me one thing, am I putting the commands at the right place or there is another console as it says 'it was valid at console'.

**Reply by Tama47** [reply]
@Albidoten Why would you struggle with MPV encoding when there are already working examples for FFmpeg? https://github.com/mpv-player/mpv/issues/9589

> mpv can encode video

Yes, that is true. But I've never seen anyone successfully encode a video with shaders applied before.

**Reply by Albidoten** [reply]
cause someone said it didn't perform well even in 12 gb vram and bcs i have intel integrated gpu so i can't even use anime4k-gui as you suggested in past.

isn't there any video explaining all that stuff, on how to do that in ffmpeg?

**Reply by Tama47** [reply]
<details>
<summary>To @Albidoten (Click on this Dropdowns for reply)</summary>

> cause someone said it didn't perform well even in 12 gb vram and bcs i have intel integrated gpu so i can't even use anime4k-gui as you suggested in past.

You might be out of luck with the FFmpeg option then (Anime4K-GUI uses FFmpeg), as it relies on Vulkan or one of the many options that require a dedicated GPU. I haven't seen anyone implement a working FFmpeg command that works without a GPU yet, but I would love to be proven wrong.

> isn't there any video explaining all that stuff, on how to do that in ffmpeg?

Nope, I wish there was too.

</details>

Your best options would be to try [Anime4KCPP](https://github.com/TianZerL/Anime4KCPP), [waifu2x](https://github.com/nagadomi/waifu2x), or [video2x](https://github.com/k4yt3x/video2x). I cannot guarantee performance, but the results should be roughly the same as Anime4K.

Or if you have access to an iOS device or a Mac, the waifu2x app on the [App Store](https://apps.apple.com/app/waifu2x/id1286485858) is by far the easiest and best way to upscale with close to real-time performance (as your device allows anyway). I'm not affiliated or sponsored in any way, but I did provide the developer with the Anime4K v4.1 shaders, so the results will be exactly the same as Anime4K.

<img width="1582" alt="image" src="https://github.com/bloc97/Anime4K/assets/96937943/a7cfb36e-76c2-42f3-8a9b-ace0579eb7cd">

**Reply by Albidoten** [reply]
@Tama47 Actually my ffmpeg and anime gui both show "libvulkan library not found" but as i've read somewhere that anime4k gui is pretty fast as it takes like 13 min for 22 min episode or something and

these three: anime4kcpp, waifu2x, video2x are too slow, i mean at that point i can use topaz instead but it was slow that's why I thought a programme who uses glsl shaders and capable of saving it internally will be a good option, but not only that with that speed with which mpv with shaders perform in realtime, it'll revolutionize the upscaling market as AI software take too much time and the outputs are comparable.

I use Windows so can't use the waifu2x Mac version and in the Windows model it's slow AF, i don't know what to do at this point, well thanks anyway for the suggestions.

**Reply by Tama47** [reply]
> Actually my ffmpeg and anime gui both show "libvulkan library not found" 

Yes, that's because you are 'missing a gpu'. I have this error on my M2 Mac too as it also have an integrated graphic. So unless you find a way to install the vulkan library somehow, which I haven't found, this method will not work.

> but as i've read somewhere that anime4k gui is pretty fast as it takes like 13 min for 22 min episode or something and

Pretty sure that would be [my comment](https://github.com/bloc97/Anime4K/discussions/236#discussioncomment-9230181). Yes, it is indeed fast, but that's because I got an RTX 3060, faster than real-time processing. Lower end graphic card would see a very different result, maybe real-time or slower.

> these three: anime4kcpp, waifu2x, video2x are too slow

I have used them in the past and they are very slow, not sure what even is the point of using Anime4K models at that point.

> I use Windows so can't use the waifu2x Mac version and in the Windows model it's slow AF, i don't know what to do at this point, well thanks anyway for the suggestions.

That's unfortunate. I've never really found a good way and convenience way to upscale on Windows either. Anime4K-GUI does works, but the video output quality isn't as high as I would like. And the FFmpeg command line is just a pain to use. On my Mac, I can just drag and drop the file to upscale and it just works. This has been my main way of upscaling whenever I need to save the video.

Also note: The windows and mac version of waifu2x is a completely different app made by different developers.

**Reply by Albidoten** [reply]
' On my Mac, I can just drag and drop the file to upscale and it just works. This has been my main way of upscaling whenever I need to save the video'   - in this line, are you talking about ffmpeg for dropping the file directly or something else?

now the main question i wanna ask is if let's say i got a medium lvl external gpu then what is the fastest way to upscale anime with the same quality as A+A mode in mpv+anime4k glsl shaders?

**Reply by Tama47** [reply]
> are you talking about ffmpeg for dropping the file directly or something else?

Waifu2x for iOS/Mac. Anime4K Mode A HQ (and all the other presets) works out of the box.

> let's say i got a medium lvl external gpu then what is the fastest way to upscale anime

Assuming you get the GPU and Vulkan API to work, fastest way ***with*** the same quality would be FFmpeg.

**Reply by Jules-A** [reply]
Probably should update that FFMPEG seems to have fixed the VRAM issue so it's working for me now.
I'm doing this: ```ffmpeg -y -i "input4.mp4" -init_hw_device vulkan -c:v hevc_amf -quality 2 -rc 0 -qp_p 19 -qp_i 19 -c:a copy -c:s mov_text -metadata:s:s:0 language=English -disposition:s:s:0 forced -vf format=yuv420p,hwupload,libplacebo=w=3840:h=2160:custom_shader_path="shaders/Myshaders.glsl",hwdownload,format=yuv420p "output4.mp4"``` Not sure how to apply multiple shaders though so I just merged them all into 1 file.
I don't use Anime4k at all though so not sure if it even works, I use a combination of modified Ravu and CfLP and a denoiser (remove .txt if interested in using): 
[Myshaders.glsl.txt](https://github.com/bloc97/Anime4K/files/15252483/Myshaders.glsl.txt)

**Reply by Albidoten** [reply]
@Jules-A  can you make a screen-recording of how you execute it as I don't understand ffmpeg that much, consider it if you got some time otherwise I'll surely try your method on my own

**Reply by Jules-A** [reply]
I just run it from a .bat file (though the full command is not general purpose since it's for AMD and forcing subtitles since my TV is dumb and doesn't play them without them being forced). Actually I just read you have integrated graphics, if Anime4k is slow, this would probably be even slower (it's just 2.6x realtime on my 6700XT, though much higher quality than Anime4k). If you wanted some GUI you could probably use Anime4k GUI and just replace the shader with one in the shader directory (rename this one to one of the names of them) of the program (never used it, just glanced code on github).

**Reply by Albidoten** [reply]
@Jules-A anime4k is not slow its just that it works in realtime but in guis due to lack of libvulkan it won't start

apart from that have you tried [this](https://github.com/haasn/libplacebo?tab=readme-ov-file), i don't know how to use it.

**Reply by Jules-A** [reply]
Oh, in that case you won't be able to use ffmpeg/libplacebo for this at all since that GUI is basically doing the same thing (sorry I didn't fully read your previous messages before responding). You can try updating drivers but with Intel you're probably out of luck.

**Reply by Albidoten** [reply]
is it possible to take 24*screenshots of mpv in 1 sec by any command? and why glsl shaders work can't be done with software like after effects?

**Reply by Tama47** [reply]
> is it possible to take 24*screenshots of mpv in 1 sec by any command?

```
Alt+s screenshot each-frame # automatically screenshot every frame; issue this command again to stop taking screenshots
```

> and why glsl shaders work can't be done with software like after effects?

GLSL shaders are designed for real-time rendering using APIs like OpenGL. However, you can convert them to WebGL to run them on the web. You could try these tools  https://free.upscaler.video/ and https://anime4k-webgpu-demo.fly.dev/

**Reply by Albidoten** [reply]
thank you to everyone who replied and finally the grind is over with the alt + s and later combining it in a video of all ss with the audio

**Reply by Tama47** [reply]
Are you just upscaling short clips? I can't see how taking screenshots and upscaling each frame, then recombining it into a video with audio could be very efficient or worth it. But assuming you really want to go this route, tools like Photoshop can export video as frame much faster than mpv `Alt+s screenshot each-frame`.

**Reply by Albidoten** [reply]
actually mpv is already upscaling it by anime4k, so at the end, i just have to combine it together. 

also, is this feature of ss of every frame is also available in media player classic.

**Reply by Tama47** [reply]
Did you do some custom configuration for mpv screenshot? Because by default, mpv saves screenshot as original frame from the video.

**Reply by Albidoten** [reply]
@Tama47 ![mpv-shot0008](https://github.com/bloc97/Anime4K/assets/159302410/05ddebeb-b89e-4c2a-bc73-f53735a5a75e)
![mpv-shot0009](https://github.com/bloc97/Anime4K/assets/159302410/92b922d4-c760-40b1-b748-3f42198695bb)
 one with and without the glsl shader on and i installed by the method, i've seen in yt, here's the [link](https://www.youtube.com/watch?v=LX2-resdeow&t=182s), and beside from that nothing extra.
 
 one extra thing i wanna add is, i installed mpv.net but since the command didn't worked in that so i installed mpv and copied all the files of mpv.net's appdata.
 
 and you did't replied on the media player classic one.
 
 edit: it is also written in manual
 
![Screenshot 2024-05-09 202427](https://github.com/bloc97/Anime4K/assets/159302410/d67143f3-86e8-49aa-9441-c260ffa8a683)

**Reply by Albidoten** [reply]
@Jules-A are your shader don't work with mpv.net?

![image](https://github.com/bloc97/Anime4K/assets/159302410/e62f90b8-9107-407b-91c2-6a00fa4ca236)

**Reply by Jules-A** [reply]
Hmm... Come to think about it, I removed the workarounds for old GPUs since I didn't plan on sharing so not sure if it works on Intel iGPUs. Also I'm pretty sure it requires using vo=gpu-next in mpv config.

**Reply by Albidoten** [reply]
@Jules-A can you share a before and after of your shader as the shader is working but not that great.

**Reply by Jules-A** [reply]
It's kind of difficult to show properly as the encode ends up ruining quality and I only have a 1440p monitor.
Anyway here's the before:
![input4](https://github.com/bloc97/Anime4K/assets/1760158/4a1b7ef6-ceb7-420f-bc8d-bdf895095e90)
and after: 
![Cells at Work! - S1E4 - Food Poisoning - 4b](https://github.com/bloc97/Anime4K/assets/1760158/6cec98a8-08be-4e28-984e-51b04875d2af)
and Anime4k A+A for comparison: 
![input4z](https://github.com/bloc97/Anime4K/assets/1760158/6b42c8eb-12b0-408a-9a3a-b463857fb141)

It's nowhere near as sharp as I mainly want to reduce aliasing and the TV applies sharpness filters over content anyway. It's also FAR, FAR less damaging to the original content despite still using denoising. On PC I use different shaders and MPV settings that end up with a sharper image though.

**Reply by Albidoten** [reply]
@Jules-A ok i am getting the same output, overall less noise, etc, so seems to be working then for me, it was suspicious as it was taking less gpu than A+A.

also, i wanna ask is there any benefits to upscaling the resolution( it seems that you've done that in the picture without AI software and can we use madvr with mpc?

**Reply by Jules-A** [reply]
Those shaders were set for 4k output so if you are using them in MPV with a lower resolution display then you likely won't see much benefit from them as they will upscale to whatever resolution you are playing at rather than 4k. Increasing resolution can greatly help with reducing aliasing, in the case of watching on 4k TV, the upscaling algorithms are far better than the TV's upscaler.
Also it's normal that it's faster than, they were encoding to 4k at 62fps vs 26fps for anime4k A+A.

**Comment by Tama47** [comment]
[Anime4K-GUI](https://github.com/mikigal/Anime4K-GUI) is an easy app with GUI you can use to save upscaled video with Anime4K shaders.
